website/docs/sql/datatype/array.md

```diff
@@ -0,0 +1,84 @@
+# Array Type
+
+An array is a sequential collection of elements packed into a single SQL value.
+Arrays are useful to model dependent information with an inherent ordering in a single row – for example, time-series data from longitudinal studies, or embedding vectors from machine learning models. 
+
+Generally, array values are denoted in curly braces with elements separated by commas, like so:
+
+```
+{"Lorem", "ipsum", "dolor", "sit", "amed"}
+```
+
+
+## Element Types and Nullability
+
+Hyper's arrays are _strongly typed_:
+All elements must be of the same type – the array's _element type_.
+The element type is a defining part of the array's overall type, meaning that `array(integer)` is a different type than `array(boolean)` or `array(text)`.
+Arrays can be built from all [atomic types](./index.md) available in Hyper.
+
+Part of an array's element type is its nullability. 
+For example, `array(smallint)` is a different array type than `array(smallint not null)`.
+Note that this is independent of the array's overall nullability.
+The following four options all represent different types in Hyper:
+
+|Type|array nullable?|elements nullable?| possible values|
+|---|---|---|---|
+|`array(integer)`|✅|✅|`{}`,`{1,2,3}`,`{1,2,null}`, `null`|
+|`array(integer not null)`|✅|❌|`{}`,`{1,2,3}`,`null`|
+|`array(integer) not null`|❌|✅|`{}`,`{1,2,3}`,`{1,2,null}`|
+|`array(integer not null) not null`|❌|❌|`{}`,`{1,2,3}`|
+
+Array types can be converted using the conventional [cast syntax](../scalar_func/conversion.md).
+
+:::info
+Non-nullable element types use less memory and enable optimizations for certain array operations. Users are therefore advised to use the most "restrictive" element type possible, given the use case at hand. 
+:::
+
+For nullable types, there exists an alternative shorthand bracket syntax of the form `type[]`. For example, `integer[]` and `array(integer)` reference the same type.
```

Comment:
Should we even mention this? If so, should we mention it in a PostgreSQL compatibility section?

Reply:
> Should we even mention this?
I think we should definitely mention it, because this is the syntax returned by, e.g., `pg_typeof`.

> If so, should we mention it in a PostgreSQL compatibility section?
We could mention that the syntax exists for compatibility reasons, but not sure if it warrants its own section. Now that `array_lower`, `array_upper`, etc being undocumented, and Hyper's arrays deviating significantly from those in PostgreSQL, what else do you suggest would go in that section?

Reply:
Do we document `pg_typeof`? If so, please then state this fact here. If not, that's yet another argument to either move this into a compatibility section or omit it add all.

I was suggesting a PostgreSQL compatibility section, because we do that in many other places of the documentation. Please skim the rest of the documentation and try to align to the general style here.

The reason why an own section is usually good is that it keeps all the "PostgreSQL compatibility cruft" out of the main reading path, so that readers who are not needing this compatibility (most of them don't) do not have to clutter their mind with this. People who just want to use arrays in Hyper do not need to be burdened with an extra syntax at all.